pkg/endpoint/policy.go

```diff
@@ -163,81 +148,151 @@ func (e *Endpoint) setNextPolicyRevision(revision uint64) {
 // however, and it is possible that policy update succeeds for some endpoints,
 // while it fails for other endpoints.
 //
-// Returns:
-//   - err: any error in obtaining information for computing policy, or if
+// Failure may be due to any error in obtaining information for computing policy,
+// or if policy could not be generated given the current set of rules in the repository.
 //
-// policy could not be generated given the current set of rules in the
-// repository.
-// Must be called with endpoint mutex held.
-func (e *Endpoint) regeneratePolicy() (retErr error) {
-	var forceRegeneration bool
+// endpoint lock must NOT be held. This is because the ipcache needs to be able to
+// make progress while generating policy, and *that* needs the endpoint unlocked to call
+// ep.ApplyPolicyMapChanges. Specifically, computing policy may cause identity allocation
+// which requires ipcache progress.
+//
+// buildMutex MUST be held, and not released until setDesiredPolicy and
+// updateRealizedState have been called
+//
+// There are a few fields that depend on this exact configuration of locking:
+//   - ep.desiredPolicy: ep.mutex must be locked between writing this and committing to
+//     the policy maps, or else policy drops may occur
+//   - ep.policyRevision: ep.mutex and ep.buildMutex must be held to write to this, or else
+//     the policyRevision may go backwards from no-op bumps from policy ingestion
+//   - ep.selectorPolicy: this may be nulled if the endpoints identity changes; we must
+//     check for this when committing. ep.mutex must be held
+//   - ep.realizedRedirects: this is read by external callers as part of policy generation,
+//     so ep.mutex must not be required to read this. Instead, both ep.mutex and ep.buildMutex
```

Comment:
I looked this up, because I was surprised. I agree that I see `DistillPolicy()` calling into this. This surprises me because distilling the policy is part of policy calculation, ie figuring out the _desired_ state. It seems odd to me for desired state calculation to take into account the realized state.

The only thing I could reason about here is that maybe the initial realization of the redirects is done through some other path, then the desired policy calculation takes this into account in order to reuse the cached port / avoid reallocating the port. Still seems somehow wrong to me, there should just be a port management facility that is able to provide that guarantee so that the policy engine doesn't have to reach into realized state to pull out the desired state.

Reply:
I believe this is used to preserve re-usable redirects across policy revisions. There also is a stateful redirect manager, I'm not sure why the endpoint also has to provide that state. @jrajahalme?

(The interesting bits are in `ep.addNewRedirectsFromDesiredPolicy()` which calls `proxy.CreateOrUpdateRedirect()`

Reply:
(to be clear, I don't expect you to address this in this PR, just raising the discussion point since it surprised me)